<div style="background:linear-gradient(135deg, rgba(249,112,102,0.18), rgba(249,112,102,0.02)); border-left:6px solid #F97066; border-radius:10px; padding:20px 24px; margin-bottom:20px;">
<h1 style="margin:0; color:#F97066; font-size:1.8em;">🎲 Teorema de Bayes</h1>
<p style="margin:6px 0 0; opacity:0.8;">Unidad 3 — Probabilidad condicional y actualización de creencias con evidencia</p>
</div>

<div style="border:1px solid rgba(14,165,233,0.35); border-radius:10px; padding:14px 20px; margin:0 0 24px; background:rgba(14,165,233,0.04);">
<strong>📑 Contenido de esta guía</strong>
<ol style="margin:8px 0 0; padding-left:20px;">
<li><a href="#que-es-bayes">¿Qué es el Teorema de Bayes?</a></li>
<li><a href="#ejercicio-1">Ejercicio 1: Helado y torta</a></li>
<li><a href="#ejercicio-2">Ejercicio 2: Empleados directivos</a></li>
<li><a href="#ejercicio-3">Ejercicio 3: Lluvia y nublado</a></li>
<li><a href="#paradoja-medica">Caso práctico: la paradoja de las pruebas médicas</a></li>
<li><a href="#cierre">Cierre y próximos pasos</a></li>
</ol>
</div>

<a id="que-es-bayes"></a>

## <span style="color:#F97066;">¿Qué es el Teorema de Bayes?</span>

El teorema de Bayes describe cómo **actualizar la probabilidad de una hipótesis a la luz de nueva evidencia**. Es la base matemática de los clasificadores *Naive Bayes* (usados, por ejemplo, en filtros de spam) y, más en general, del razonamiento bayesiano: partir de una creencia inicial e ir refinándola a medida que se observan nuevos datos.

$$P(A|B) = \frac{P(A) \cdot P(B|A)}{P(B)}$$

<div style="border-radius:12px; overflow:hidden; border:1px solid rgba(14,165,233,0.35); margin:16px 0;">
<table style="width:100%; border-collapse:collapse; font-size:0.95em;">
<thead>
<tr style="background:#0EA5E9;">
<th style="text-align:left; padding:10px 14px; color:#ffffff;">Término</th>
<th style="text-align:left; padding:10px 14px; color:#ffffff;">Nombre</th>
<th style="text-align:left; padding:10px 14px; color:#ffffff;">Significado</th>
</tr>
</thead>
<tbody>
<tr style="background:rgba(14,165,233,0.07);">
<td style="padding:10px 14px;"><code>P(A)</code></td>
<td style="padding:10px 14px;">Prior (probabilidad previa)</td>
<td style="padding:10px 14px;">Qué tan probable se creía que era A <em>antes</em> de observar la evidencia B.</td>
</tr>
<tr>
<td style="padding:10px 14px;"><code>P(B|A)</code></td>
<td style="padding:10px 14px;">Verosimilitud (<em>likelihood</em>)</td>
<td style="padding:10px 14px;">Qué tan probable es observar la evidencia B, si A fuera cierto.</td>
</tr>
<tr style="background:rgba(14,165,233,0.07);">
<td style="padding:10px 14px;"><code>P(B)</code></td>
<td style="padding:10px 14px;">Evidencia</td>
<td style="padding:10px 14px;">Qué tan probable es observar B, considerando todos los casos posibles.</td>
</tr>
<tr>
<td style="padding:10px 14px;"><code>P(A|B)</code></td>
<td style="padding:10px 14px;">Posterior</td>
<td style="padding:10px 14px;">Qué tan probable es A <em>después</em> de tener en cuenta la evidencia B.</td>
</tr>
</tbody>
</table>
</div>

A continuación, algunos ejercicios para practicar su aplicación.

<a id="ejercicio-1"></a>

## <span style="color:#F97066;">Ejercicio 1</span>

<span style="background:#0EA5E9;color:#ffffff;padding:4px 14px;border-radius:14px;font-size:0.75em;">🧪 EJERCICIO</span>

En una academia, la probabilidad de que a un alumno seleccionado al azar le guste el helado es del 60 %, mientras que la probabilidad de que a un alumno le guste la torta es del 36 %. Además, se sabe que la probabilidad de que a un alumno le guste la torta dado que le gusta el helado es del 40 %. Calcular la probabilidad de que a un alumno le guste el helado, dado que le gusta la torta.

**Teorema de Bayes:**

$P(A|B)=\frac{P(A)P(B|A)}{P(B)}$

In [1]:
# Probabilidades proporcionadas
P_helado = 0.60
P_torta = 0.36
P_torta_dado_helado = 0.40

# Aplicar el teorema de Bayes para calcular la probabilidad de gustar helado dado que le gusta la torta
P_helado_dado_torta = (P_torta_dado_helado * P_helado) / P_torta

# Imprimir el resultado
print(
    f'''La probabilidad de que a un alumno le guste el helado dado que le gusta la torta es:
    {round(P_helado_dado_torta, 3)*100}%'''
)

La probabilidad de que a un alumno le guste el helado dado que le gusta la torta es:
    66.7%


<a id="ejercicio-2"></a>

## <span style="color:#F97066;">Ejercicio 2</span>

<span style="background:#0EA5E9;color:#ffffff;padding:4px 14px;border-radius:14px;font-size:0.75em;">🧪 EJERCICIO</span>

El 20 % de los empleados de una empresa son ingenieros y otro 20 % son economistas. El 75 % de los ingenieros ocupan un puesto directivo y el 50 % de los economistas también, mientras que los no ingenieros y los no economistas solamente el 20 % ocupa un puesto directivo. ¿Cuál es la probabilidad de que un empleado directivo elegido al azar sea ingeniero?

<div style="border-left:4px solid #6366F1; background:rgba(99,102,241,0.08); border-radius:6px; padding:10px 16px; margin:16px 0;">
<strong>📝 Nota</strong><br>
Este ejercicio requiere calcular primero <code>P(directivo)</code> mediante la <strong>ley de probabilidad total</strong>: como todo empleado es ingeniero, economista, o ninguno de los dos (tres grupos que no se solapan y cubren a todos los empleados), <code>P(directivo)</code> es la suma ponderada de la probabilidad de ser directivo en cada uno de esos grupos.
</div>

In [2]:
# Probabilidades proporcionadas
P_ing = 0.20
P_eco = 0.20
P_no_ing_no_eco = 0.60
P_dir_dado_ing = 0.75
P_dir_dado_eco = 0.50
P_dir_dado_no_ing_no_eco = 0.20

# Calcular la probabilidad de que sea directivo
P_dir = (
    (P_dir_dado_ing * P_ing)
    + (P_dir_dado_eco * P_eco)
    + (P_dir_dado_no_ing_no_eco * P_no_ing_no_eco)
)

# Calcular la probabilidad de que un empleado directivo sea ingeniero
P_ing_dado_dir = (P_dir_dado_ing * P_ing) / (P_dir)

# Imprimir el resultado
print(
    "La probabilidad de que un empleado directivo elegido al azar sea ingeniero es:",
    round(P_ing_dado_dir, 3) * 100,
    "%",
)

La probabilidad de que un empleado directivo elegido al azar sea ingeniero es: 40.5 %


<a id="ejercicio-3"></a>

## <span style="color:#F97066;">Ejercicio 3</span>

<span style="background:#0EA5E9;color:#ffffff;padding:4px 14px;border-radius:14px;font-size:0.75em;">🧪 EJERCICIO</span>

Suponer que se tienen las siguientes probabilidades:

- $P(lluvia) = 0.20$
- $P(nublado) = 0.40$
- $P(nublado|lluvia) = 0.85.$

Calcular $P(lluvia|nublado)$. A diferencia de los ejercicios anteriores, aquí se define una función reutilizable que implementa directamente la fórmula del teorema de Bayes.

In [3]:
# Definición de función de Bayes
def bayesTheorem(pA, pB, pBA):
    return (pA * pBA) / pB

# Definición de probabilidades
pLluvia = 0.20
pNublado = 0.40
pNubladoDadoLluvia = 0.85

# Calcular P(lluvia|nublado) con la función
print(
    f"La probabilidad de que llueva si está nublado es: {round(bayesTheorem(pLluvia, pNublado, pNubladoDadoLluvia) *100, 3)}%"
)

La probabilidad de que llueva si está nublado es: 42.5%


---

<a id="paradoja-medica"></a>

# <span style="color:#F97066;">Caso práctico: la paradoja de las pruebas médicas</span>

Uno de los usos más conocidos (y más contraintuitivos) del teorema de Bayes es interpretar correctamente el resultado de una prueba diagnóstica. Considere una enfermedad poco común y una prueba médica con las siguientes características:

- **Prevalencia**: solo el **1 %** de la población tiene la enfermedad: $P(enfermo) = 0.01$.
- **Sensibilidad**: si una persona está enferma, la prueba da positivo el **99 %** de las veces: $P(positivo|enfermo) = 0.99$.
- **Especificidad**: si una persona está sana, la prueba da negativo el **95 %** de las veces (es decir, un 5 % de falsos positivos): $P(positivo|sano) = 0.05$.

**Pregunta:** si una persona da positivo en la prueba, ¿cuál es la probabilidad de que realmente tenga la enfermedad, $P(enfermo|positivo)$?

<span style="background:#0EA5E9;color:#ffffff;padding:4px 14px;border-radius:14px;font-size:0.75em;">🧪 EJERCICIO</span>

Antes de ejecutar la siguiente celda, intente estimar la respuesta intuitivamente. La mayoría de las personas responde "alrededor del 90-95 %" (confundiendo la especificidad de la prueba con la probabilidad buscada). El resultado real suele sorprender.

In [4]:
# Probabilidades del caso
p_enfermo = 0.01                       # Prevalencia: prior P(enfermo)
p_positivo_dado_enfermo = 0.99         # Sensibilidad: P(positivo | enfermo)
p_positivo_dado_sano = 0.05            # 1 - especificidad: P(positivo | sano)
p_sano = 1 - p_enfermo

# Ley de probabilidad total: P(positivo) considerando ambos grupos (enfermos y sanos)
p_positivo = (p_positivo_dado_enfermo * p_enfermo) + (p_positivo_dado_sano * p_sano)

# Teorema de Bayes: P(enfermo | positivo)
p_enfermo_dado_positivo = (p_positivo_dado_enfermo * p_enfermo) / p_positivo

print(f"P(positivo) = {round(p_positivo, 4)}")
print(f"P(enfermo | positivo) = {round(p_enfermo_dado_positivo * 100, 1)}%")

P(positivo) = 0.0594
P(enfermo | positivo) = 16.7%


A pesar de que la prueba es bastante precisa (99 % de sensibilidad, 95 % de especificidad), la probabilidad real de estar enfermo tras un resultado positivo es de apenas **16.7 %**, no 90-95 %. Esto ocurre porque la enfermedad es muy poco frecuente (1 % de prevalencia): la mayoría de los resultados positivos provienen, en términos absolutos, del enorme grupo de personas sanas (falsos positivos), y no del pequeño grupo de personas realmente enfermas.

Esta es la razón por la cual, en diagnóstico médico, en detección de fraude o en filtros de spam, **la prevalencia del evento que se busca detectar es tan importante como la precisión del propio clasificador**: un modelo o prueba con buena sensibilidad y especificidad puede seguir siendo poco útil si el evento que detecta es muy raro.

---

<a id="cierre"></a>

# <span style="color:#F97066;">🎯 Cierre y próximos pasos</span>

<div style="border-left:4px solid #14B8A6; background:rgba(20,184,166,0.08); border-radius:6px; padding:10px 16px; margin:16px 0;">
<strong>✅ Resumen</strong><br>
En este notebook se presentó el teorema de Bayes:

- La fórmula $P(A|B) = \frac{P(A) \cdot P(B|A)}{P(B)}$ y el significado de cada uno de sus términos: prior, verosimilitud, evidencia y posterior.
- La ley de probabilidad total, para calcular <code>P(B)</code> cuando no se conoce directamente.
- Tres ejercicios de aplicación directa de la fórmula.
- La paradoja de las pruebas médicas: por qué la prevalencia de un evento raro afecta drásticamente la interpretación de un resultado positivo, incluso con clasificadores muy precisos.

</div>

<div style="border:1px solid rgba(14,165,233,0.35); border-radius:10px; padding:14px 20px; margin:16px 0; background:rgba(14,165,233,0.04);">
<strong>➡️ Continúe con</strong>
<ul style="margin:8px 0 0; padding-left:20px;">
<li><code>4-análisis-exploratorio-datos-EDA.ipynb</code> — el primer paso práctico antes de entrenar cualquier modelo: explorar y comprender los datos.</li>
</ul>
</div>